### ISSUE TITLE
bad explicitouter/erasure interaction regarding transform of isInstanceOf[Outer.this.Module.type] to eq(Outer.this.Module)

### ISSUE THREAD

**Issue by scabug** [scala/bug#7598]
```scala
package com.middlewareman

trait Component

case class Event(component: Component)

class Owner {

  object Timestamp extends Component {
    val state = System.currentTimeMillis()
    def unapply(name: Timestamp.type) = Some(name.state)
  }

  val handler: PartialFunction[Event, Unit] = {
    case Event(Timestamp(time)) => println(time)
  }

}
```
I have not tried compiling outside Eclipse.

**Comment by SethTisue** [comment]
Somehow, some way, got fixed in 2.13.6 – no compiler crash. Remains fixed in 3.2.2.

**Comment by som-snytt** [comment]
https://github.com/scala/scala/pull/9504 perchance
